### PR TITLE
Nguyenquyhy/17 custom linear gauges

### PR DESCRIPTION
**Pull request by rmroc451**
Hopefully this works more for issue #17.  I'll be on PTO this week and maybe tinkering with spherical custom gauges.

### CODE REVIEW

File: FlightStreamDeck.Logics/ImageLogic.cs [hunk 1]
```diff
@@ -231,5 +320,43 @@ public string GetGaugeImage(string text, float value, float min, float max)
 
             return "data:image/png;base64, " + base64;
         }
+
+        private void setupValue(bool top, string labelText, string value, float ratio, int img_width, float chevronSize, 
+            float width_margin, float chart_width, float max, IImageProcessingContext ctx
+        )
+        {
+            if (labelText.Length > 0)
```
Comment (sn3akz) [inline FlightStreamDeck.Logics/ImageLogic.cs:328]:
```
if (labelText.Length == 0) 
{
    return;
} 
```

Would save you a bunch of indents :)

File: FlightStreamDeck.Logics/ImageLogic.cs [hunk 2]
```diff
@@ -217,12 +225,93 @@ public string GetGaugeImage(string text, float value, float min, float max)
...
-                    ctx.DrawText(text, titleFont, Color.White, new PointF(HALF_WIDTH - size.Width / 2, 57));
+                    ctx.DrawText(text, titleFont, Color.White, new PointF(HALF_WIDTH - size.Width / 3, 46));
                 }
 
                 var valueText = value.ToString();
                 Color textColor = value > max ? Color.Red : Color.White;
-                ctx.DrawText(valueText, font, textColor, new PointF(25, 30));
+                ctx.DrawText(valueText, font, textColor, new PointF(18, 20));
+            });
+
```
Comment (sn3akz) [inline FlightStreamDeck.Logics/ImageLogic.cs:235]: extraneous empty lines

File: FlightStreamDeck.Logics/ImageLogic.cs [hunk 3]
```diff
@@ -15,7 +16,12 @@ public interface IImageLogic
         string GetNumberImage(int number);
         string GetNavComImage(string type, bool dependant, string value1 = null, string value2 = null, bool showMainOnly = false);
         public string GetHorizonImage(float pitchInDegrees, float rollInDegrees, float headingInDegrees);
-        public string GetGaugeImage(string text, float value, float min, float max);
+        public string GetGenericGaugeImage(string text, float value, float min, float max);
+        public string GetCustomGaugeImage(string textTop, string textBottom,
```
Comment (sn3akz) [inline FlightStreamDeck.Logics/ImageLogic.cs:20]: Maybe either 1 variable for each line, or all variables on 1 line, but grouped by two I haven't seen before :)

File: FlightStreamDeck.Logics/Actions/CustomGaugeAction.cs [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+        private readonly EnumConverter enumConverter;
+
+        private TOGGLE_EVENT? toggleEvent = null;
+        private TOGGLE_VALUE? displayValueTop = null;
+        private TOGGLE_VALUE? displayValueBottom = null;
+
+        private string currentValueTop = "0";
+        private string currentValueBottom = "0";
+
+        public CustomGaugeSettings settings = new CustomGaugeSettings() {
```
Comment (sn3akz) [inline FlightStreamDeck.Logics/Actions/CustomGaugeAction.cs:57]: Inconsitent { placement. Do we { on newlines or not @nguyenquyhy ?

Comment (nguyenquyhy) [reply FlightStreamDeck.Logics/Actions/CustomGaugeAction.cs:57]: I would prefer { on new line (unless everything on the same line) as my VS will auto format like that anyway.

File: FlightStreamDeck.Logics/ImageLogic.cs (hunk 3 above)
Comment (nguyenquyhy) [reply FlightStreamDeck.Logics/ImageLogic.cs:20]: I usually put one line per group of parameters also :P.

File: FlightStreamDeck.Logics/ImageLogic.cs (hunk 2 above)
Comment (nguyenquyhy) [reply FlightStreamDeck.Logics/ImageLogic.cs:235]: Agree with this. Should reserve double new line when you really want to create a "special" separation.

File: FlightStreamDeck.Logics/ImageLogic.cs (hunk 1 above)
Comment (nguyenquyhy) [reply FlightStreamDeck.Logics/ImageLogic.cs:328]: Yes I would prefer "return" style on top for parameter validation. Also, for string, string.IsNullOrWhiteSpace would normally catch all scenarios that you want to skip.

**Comment by nguyenquyhy** [comment]
Is there a reason for the button to be a new one instead of enhancement to the current gauge button?
If the new button is really existing one + new feature without losing existing functionality, I would really prefer to have just 1 button.

**Review by nguyenquyhy — COMMENTED** [review]
This commit is marked as WIP. Do I need to worry about anything here?

**Review by nguyenquyhy — COMMENTED** [review]
Is there a reason for the button to be a new one instead of enhancement to the current gauge button?
If the new button is really existing one + new feature without losing existing functionality, I would really prefer to have just 1 button.

**Comment by rmroc451** [comment]
> This commit is marked as WIP. Do I need to worry about anything here?

It was wip when I started, I removed that wip block.

**Comment by rmroc451** [comment]
> Is there a reason for the button to be a new one instead of enhancement to the current gauge button?
> If the new button is really existing one + new feature without losing existing functionality, I would really prefer to have just 1 button.

Didn't know how you'd prefer it. If we end up liking it, we can decom the old one.

I didn't want to muddle that generic gauge class.

File: FlightStreamDeck.Logics/Actions/CustomGaugeAction.cs (hunk 4 above)
Comment (rmroc451) [reply FlightStreamDeck.Logics/Actions/CustomGaugeAction.cs:57]: Got it, Its visual flavor. But I'll conform. :)

File: FlightStreamDeck.Logics/ImageLogic.cs (hunk 1 above)
Comment (rmroc451) [reply FlightStreamDeck.Logics/ImageLogic.cs:328]: I'm of the camp a function should only have one return statement, period.

If y'all want me to short circuit stuff, I'll try to overcome that habit.

**Comment by nguyenquyhy** [comment]
> Didn't know how you'd prefer it. If we end up liking it, we can decom the old one.
> 
> I didn't want to muddle that generic gauge class.

I like it, but I at the same time I want to avoid changing button ID (by decom the current gauge) since that would break existing profile. If there is some design philosophy difference, it would be ok to keep both. If one is an extension of the other, I would prefer merging them.

I think I'll take a look and try to merge them before release.

**Comment by rmroc451** [comment]
I'll merge them in the next day or so :)

No worries!

File: FlightStreamDeck.Logics/Actions/CustomGaugeAction.cs (hunk 4 above)
Comment (sn3akz) [reply FlightStreamDeck.Logics/Actions/CustomGaugeAction.cs:57]: True, it's all visual flavor. Same for the parameter grouping. But having a consistent choice (and I don't mind too much which choice) makes the code more readable and tidy :)

**Comment by rmroc451** [comment]
closing for now, I'll reopen the PR when I get the branch sorted out.

**Pull request closed without merging**